webdriver/support/fixtures.py

```diff
@@ -83,54 +88,88 @@ def create_window():
         return new_windows.pop()
     return create_window
 
+
 def http(session):
     return HTTPRequest(session.transport.host, session.transport.port)
 
+
 def server_config():
     return json.loads(os.environ.get("WD_SERVER_CONFIG"))
 
-def create_session(request):
-    """Provide a factory function that produces wdclient `Session` instances.
-    If the `WD_CAPABILITIES` environment variable is set, it will be parsed as
-    JSON and the resulting object will be included in the WebDriver "Create
-    Session" command. Additional capabilities may be specified as an optional
-    argument to this function, but the operation will fail if any values
-    conflict with those specified via the environment. If the session is still
-    active at the completion of the test, it will be destroyed
-    automatically."""
-
-    def create_session(test_capabilities=None):
-        host = os.environ.get("WD_HOST", default_host)
-        port = int(os.environ.get("WD_PORT", default_port))
-        if test_capabilities is None:
-            test_capabilities = {}
-        env_capabilities = json.loads(os.environ.get("WD_CAPABILITIES", "{}"))
-
-        capabilities = merge_dictionaries(env_capabilities, test_capabilities)
-        session = webdriver.Session(host, port, capabilities=capabilities)
-
-        def destroy():
-            if session.session_id is not None:
-                session.end()
-
-        # finalisers are popped off a stack, making their ordering reverse
-        request.addfinalizer(destroy)
-        request.addfinalizer(lambda: _switch_to_top_level_browsing_context(session))
-        request.addfinalizer(lambda: _restore_windows(session))
-        request.addfinalizer(lambda: _dismiss_user_prompts(session))
-        request.addfinalizer(lambda: _ensure_valid_window(session))
-
-        return session
+
+def configuration():
+    host = os.environ.get("WD_HOST", default_host)
+    port = int(os.environ.get("WD_PORT", default_port))
+    capabilities = json.loads(os.environ.get("WD_CAPABILITIES", "{}"))
+
+    return {
+        "host": host,
+        "port": port,
+        "capabilities": capabilities
+    }
+
+
+_current_session = None
+
+
+def session(configuration, request):
+    """Create and start a session for a test that does not itself test session creation.
+
+    By default the session will stay open after each test, but we always try to start a
+    new one and assume that if that fails there is already a valid session. This makes it
+    possible to recover from some errors that might leave the session in a bad state, but
+    does not demand that we start a new session per test."""
```

Comment:
Not entirely sure this is true --- if the session still claims to be active then we're in trouble.